File: golang/743_network_delay_time/network_delay_time.go
// Package main ...
package main

import (
	"math"
	"testing"

	"github.com/stretchr/testify/assert"
)

func networkDelayTime_Dijkstra(times [][]int, n int, k int) int {
	arcs := make([][]int, n+1)
	for i := 0; i < len(arcs); i++ {
		arcs[i] = make([]int, n+1)
		for j := 0; j < len(arcs[i]); j++ {
			arcs[i][j] = 10010
		}
	}
	for i := 1; i <= n; i++ {
		arcs[i][i] = 0
	}

	for _, time := range times {
		arcs[time[0]][time[1]] = time[2]
	}

	dist := arcs[k]
	flag := make([]bool, n+1)
	for i := 0; i < len(flag); i++ {
		flag[i] = false
	}
	flag[k] = true

	for i := 1; i <= n; i++ {
		minn := 10010
		pos := 0
		for j := 1; j <= n; j++ {
			if !flag[j] && dist[j] < minn {
				pos = j
				minn = dist[j]
			}
		}
		flag[pos] = true
		for j := 1; j <= n; j++ {
			if !flag[j] && dist[pos]+arcs[pos][j] < dist[j] {
				dist[j] = dist[pos] + arcs[pos][j]
			}
		}
	}
	ret := math.MinInt32
	for i := 1; i < len(arcs[k]); i++ {
		if arcs[k][i] > ret {
			ret = arcs[k][i]
		}
	}
	if ret == 10010 {
		return -1
	}
	return ret
}

func networkDelayTime_floyd(times [][]int, n int, k int) int {
	arcs := make([][]int, n+1)
	for i := 0; i < len(arcs); i++ {
		arcs[i] = make([]int, n+1)
		for j := 0; j < len(arcs[i]); j++ {
			arcs[i][j] = 10010
		}
	}
	for i := 1; i <= n; i++ {
		arcs[i][i] = 0
	}

	for _, time := range times {
		arcs[time[0]][time[1]] = time[2]
	}

	for r := 1; r <= n; r++ {
		for i := 1; i <= n; i++ {
			for j := 1; j <= n; j++ {
				if arcs[i][j] > arcs[i][r]+arcs[r][j] {
					arcs[i][j] = arcs[i][r] + arcs[r][j]
				}
			}
		}
	}
	ret := math.MinInt32
	for i := 1; i < len(arcs[k]); i++ {
		if arcs[k][i] > ret {
			ret = arcs[k][i]
		}
	}
	if ret == 10010 {
		return -1
	}
	return ret
}

func main() {
	tests := []struct {
		times [][]int
		n     int
		k     int
		ans   int
	}{
		{[][]int{{2, 1, 1}, {2, 3, 1}, {3, 4, 1}}, 4, 2, 2},
		{[][]int{{1, 2, 1}}, 2, 1, 1},
		{[][]int{{1, 2, 1}}, 2, 2, -1},
	}

	for index, test := range tests {
		assert.Equal(&testing.T{}, test.ans, networkDelayTime_Dijkstra(test.times, test.n, test.k), index)
		assert.Equal(&testing.T{}, test.ans, networkDelayTime_floyd(test.times, test.n, test.k), index)
	}
}

File: rust/743_network_delay_time/network_delay_time.rs
struct Solution;

impl Solution {
    pub fn network_delay_time_dijkstra(times: Vec<Vec<i32>>, n: i32, k: i32) -> i32 {
        let n = n as usize;
        let k = k as usize;
        let mut arcs: Vec<Vec<i32>> = vec![vec![10001; n + 1]; n + 1];
        for i in 1..=n {
            arcs[i][i] = 0;
        }

        for time in times.iter() {
            arcs[time[0] as usize][time[1] as usize] = time[2];
        }
        let mut dist = arcs[k].clone();
        let mut flag = vec![false; n + 1];
        flag[k] = true;
        for _ in 1..=n {
            let mut minn = 10010;
            let mut pos = 0;
            for j in 1..=n {
                if !flag[j] && dist[j] < minn {
                    pos = j;
                    minn = dist[j];
                }
            }
            flag[pos] = true;
            for j in 1..=n {
                if !flag[j] && dist[pos] + arcs[pos][j] < dist[j] {
                    dist[j] = dist[pos] + arcs[pos][j];
                }
            }
        }
        let mut ret = i32::MIN;
        for i in 1..dist.len() {
            if dist[i] > ret {
                ret = dist[i]
            }
        }
        match ret == 10001 {
            true => -1,
            false => ret,
        }
    }

    pub fn network_delay_time_floyd(times: Vec<Vec<i32>>, n: i32, k: i32) -> i32 {
        let n = n as usize;
        let k = k as usize;
        let mut arcs: Vec<Vec<i32>> = vec![vec![10010; n + 1]; n + 1];
        for i in 1..=n {
            arcs[i][i] = 0;
        }

        for time in times.iter() {
            arcs[time[0] as usize][time[1] as usize] = time[2];
        }

        for r in 1..=n {
            for i in 1..=n {
                for j in 1..=n {
                    if arcs[i][j] > arcs[i][r] + arcs[r][j] {
                        arcs[i][j] = arcs[i][r] + arcs[r][j];
                    }
                }
            }
        }
        let mut ret = i32::MIN;
        for i in 1..arcs[k].len() {
            if arcs[k][i] > ret {
                ret = arcs[k][i]
            }
        }
        match ret == 10010 {
            true => -1,
            false => ret,
        }
    }
}

fn main() {
    let tests = vec![
        (vec![vec![2, 1, 1], vec![2, 3, 1], vec![3, 4, 1]], 4, 2, 2),
        (vec![vec![1, 2, 1]], 2, 1, 1),
        (vec![vec![1, 2, 1]], 2, 2, -1),
    ];

    for (times, n, k, ans) in tests {
        assert_eq!(
            Solution::network_delay_time_dijkstra(times.clone(), n, k),
            ans
        );
        assert_eq!(Solution::network_delay_time_floyd(times.clone(), n, k), ans);
    }
}
